stdlib/public/core/UnicodeHelpers.swift

```diff
@@ -64,7 +64,7 @@ internal func _decodeUTF8(
 internal func _decodeScalar(
   _ utf16: UnsafeBufferPointer<UInt16>, startingAt i: Int
 ) -> (Unicode.Scalar, scalarLength: Int) {
-  let high = utf16[i]
+  let high = utf16[_unchecked: i]
   if i + 1 >= utf16.count {
```

Comment:
If this was:

```swift
guard i < utf16.endIndex else { ... }
```

Then the index for `low` could be calculated without trapping on overflow (`i &+ 1`). I think, as it is written, the compiler must _always_ evaluate `i + 1` and check for overflow.

Reply:
I don't think llvm has much trouble figuring out that `i < utf16.count` here. (So e.g. switching to the unchecked subscript probably changes nothing for the i+1 subscript invocation below.)

One way to verify this is to look at the generated code to see if we actually have an overflow branch for the addition. If we do, then switching to wrapping arithmetic could be worth investigating in a followup PR!

Reply:
Could it be that this function is not even used?

Reply:
Quite possible! These functions have a precondition that the buffer they're given contains valid UTF-8/UTF-16 data. This precludes their use with NSString, and we no longer have a native string form that uses UTF-16.